### ISSUE TITLE
false positive: relative-beyond-top-level with namespace modules when specific file is processed in parallel mode

### ISSUE THREAD

**Issue by felixfontein** [pylint-dev/pylint#10794]
### Bug description

Files with content:

```
a/b.py      ->      B = 1
a/c/d.py    ->      from ..b import B
```

If I run one of:
```
pylint --jobs 0 --disable=all --enable=relative-beyond-top-level --source-roots . a/
pylint --jobs 1 --disable=all --enable=relative-beyond-top-level --source-roots . a/c/d.py
```
everything is well. But if I run:
```
pylint --jobs 0 --disable=all --enable=relative-beyond-top-level --source-roots . a/c/d.py
```
I get an error:
```
************* Module d
a/c/d.py:1:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
```

(The only potentially related bug I could find is #10147, but I'm not sure whether it's really related.)

### Configuration

```ini

```

### Command used

```shell
mkdir -p a/c
echo "B = 1" > a/b.py
echo "from ..b import B" > a/c/d.py
# These ones are successful:
pylint --jobs 0 --disable=all --enable=relative-beyond-top-level --source-roots . a/
pylint --jobs 1 --disable=all --enable=relative-beyond-top-level --source-roots . a/c/d.py
# This one is not, and produces the output shown below:
pylint --jobs 0 --disable=all --enable=relative-beyond-top-level --source-roots . a/c/d.py
```

### Pylint output

```python
************* Module d
a/c/d.py:1:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

pylint should be successful in both calls.

### Pylint version

```shell
pylint 4.0.4
astroid 4.0.2
Python 3.13.11 (main, Dec  7 2025, 13:01:45) [GCC 15.2.1 20251112]
```

### Additional dependencies

```python

```

**Comment by felixfontein** [comment]
CC @webknjaz - this might be related to your #10147, though this one definitely shouldn't be caused by setting the source root in a wrong way.

**Comment by felixfontein** [comment]
@DanielNoord I installed the latest pytest and astroid from `main` branch, and this error still happens. It does not happen in the tests, but if I run the test commands manually:
```
. .tox/py314/bin/activate
cd tests/regrtest_data/pep420/issue_10794
pylint --rcfile=../../../../.tox/py314/lib/python3.14/site-packages/pylint/testutils/testing_pylintrc --jobs 0 --disable=all --enable=relative-beyond-top-level --source-roots . a/c/d.py
```
I still get this error:
```
************* Module d
a/c/d.py:1:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```
Did you try running this *outside* of pytest, or just as part of your reproducer you added in #10889?

From my tries to debug this (I tried it multiple times last year), this seems to be some interaction between Python's module loader and astroid and pylint, and I could imagine that inside pytest this behaves different (so the test you added works fine).

**Comment by felixfontein** [comment]
Sorry, I meant to ping @Julfried...

**Comment by felixfontein** [comment]
When I do add the first commit from #10889, namely https://github.com/pylint-dev/pylint/pull/10889/changes/f82dcad9e1eb6f3a67bb5254d07aecee6b652e76, this actually fixes the problem for me.

**Comment by felixfontein** [comment]
(I'm testing this with Python 3.14 on Arch Linux.)

**Comment by Julfried** [comment]
Very strange, I am unable to reproduce this on Fedora 43 using the current main branch. I am not using tox though, however I would not expect that alone to change the behavior here.

Command I used:
```bash
pylint --rcfile=pylint/testutils/testing_pylintrc --jobs 0 --disable=all --enable=relative-beyond-top-level --source-roots . a/c/d.py
```
Output I get:

```
--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

pylint version:
```
pylint 4.1.0-dev0
astroid 4.1.1
Python 3.14.3 (main, Feb  4 2026, 00:00:00) [GCC 15.2.1 20260123 (Red Hat 15.2.1-7)]
```

**Comment by felixfontein** [comment]
I only used tox's venv to re-use the virtual environment in which the unit tests passed, to see whether it's caused by that. I also tried stripping down the unit tests in pylint to only run that single test that fails for me outside of pytest, and it didn't fail either. So my guess is that pytest does some magic which prevents the failure.

I can reproduce it with `docker run --rm -it archlinux:latest`, and then:
```
pacman -Sy
pacman -S --noconfirm python python-pip
pip install --break-system-packages https://github.com/pylint-dev/pylint/archive/main.tar.gz https://github.com/pylint-dev/astroid/archive/main.tar.gz 
mkdir -p a/c
echo "B = 1" > a/b.py
echo "from ..b import B" > a/c/d.py
pylint --jobs 0 --disable=all --enable=relative-beyond-top-level --source-roots . a/c/d.py
```
I can also reproduce it with `docker run --rm -it fedora:latest` (that's Fedora 43):
```
dnf install -y python-pip
pip install --break-system-packages https://github.com/pylint-dev/pylint/archive/main.tar.gz https://github.com/pylint-dev/astroid/archive/main.tar.gz 
mkdir -p a/c
echo "B = 1" > a/b.py
echo "from ..b import B" > a/c/d.py
pylint --jobs 0 --disable=all --enable=relative-beyond-top-level --source-roots . a/c/d.py
```

**Comment by Julfried** [comment]
I am also unable to reproduce this in docker (neither Arch nor Fedora). Maybe someone else could try? @Pierre-Sassoulas 

<img width="1910" height="957" alt="Image" src="https://github.com/user-attachments/assets/309d8a34-7f64-4921-9e76-2f1b286a6fda" />

**Comment by Pierre-Sassoulas** [comment]
```dockerfile
FROM python:3.13-slim

COPY . /pylint
RUN pip install /pylint

RUN mkdir -p /app/a/c
RUN echo "B = 1" > /app/a/b.py
RUN echo "from ..b import B" > /app/a/c/d.py

WORKDIR /app

RUN cat <<'SCRIPT' > /app/test.sh
#!/bin/bash
failed=0
run_test() {
    local label="$1"; shift
    echo "=== $label ==="
    if "$@"; then
        echo "PASS"
    else
        echo "FAIL (exit $?)"
        failed=1
    fi
    echo
}
run_test "directory, parallel" \
    pylint --jobs 0 --disable=all --enable=relative-beyond-top-level --source-roots . a/
run_test "single file, serial" \
    pylint --jobs 1 --disable=all --enable=relative-beyond-top-level --source-roots . a/c/d.py
run_test "single file, parallel (BUG)" \
    pylint --jobs 0 --disable=all --enable=relative-beyond-top-level --source-roots . a/c/d.py
exit $failed
SCRIPT
RUN chmod +x /app/test.sh

CMD ["/app/test.sh"]

```
This fail for me locally either with pylint 4.0.4 or pylint current main :
```
=== directory, parallel ===                                                                                                                                                                                                

------------------------------------
Your code has been rated at 10.00/10

PASS

=== single file, serial ===

------------------------------------
Your code has been rated at 10.00/10

PASS

=== single file, parallel (BUG) ===
************* Module d
a/c/d.py:1:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)

-----------------------------------
Your code has been rated at 0.00/10

FAIL (exit 2)
```

**Comment by Julfried** [comment]
Ok strange if I install pylint with `pip install -e .` everything works, however when installed via `pip install .` I am able to reproduce this. Will work on this.

**Comment by Julfried** [comment]
Could you check my [branch](https://github.com/Julfried/pylint/tree/fix-parallel)? For me it is now fixed.

**Comment by felixfontein** [comment]
@Julfried your branch fixes the problem for me!

**Comment by Julfried** [comment]
Nice, to hear. Then I will do a PR for that :)

**Comment by Julfried** [comment]
Unfortunately I had to change some stuff to make tests pass. Just to be sure, could you please verify that is is still fixed in [here](https://github.com/Julfried/pylint/tree/fix-parallel)?

**Comment by felixfontein** [comment]
@Julfried the current version of your branch still fixes the issue for me!
